src/main/scala/fs2/async/mutable/Queue.scala

```diff
@@ -0,0 +1,282 @@
+              // Results in tuple containing accumulated `A` with size <= require,  and enqueuers that were not used yet if any
+              // may return both empty, in that case queue is empty.
+              def satisfyDequeue(require:Int, enq:Vector[EnQueue[F,A]],acc:Vector[A]): F[(Vector[A],Vector[EnQueue[F,A]])] = {
+                if (acc.size >= require) F.pure(acc -> enq)
+                else {
+                  enq.headOption match {
+                    case None => F.pure(acc -> Vector.empty)
+                    case Some((xa,enqF)) =>
+                      val (nacc, rem) = (acc ++ xa).splitAt(require)
+                      if (rem.isEmpty) F.bind(enqF(0))(_ => satisfyDequeue(require,enq.tail, nacc))
+                      else F.pure((nacc,(rem,enqF) +: enq.tail))
+                  }
+                }
+              }
+
+              // Signals size change of queue, if that has changed
+              def signalSize(s:State[F,A], ns:State[F,A]) : F[Unit] = {
+                if (s.queue.size != ns.queue.size) szSignal.set(ns.queue.size)
+                else F.pure(())
+              }
+
+              // enqueueAll implementation
+              // tries to satisfy or waiting dequeuers on this enqueue and/or attaches the remainder to the queue
+              // also updates size of the queue.
+              def _enqueueAll(xa: Seq[A]): F[Int] = {
+                if (xa.isEmpty) F.map(F.get(qref))(_.queue.size)
+                else {
+                  F.bind(F.ref[Int]) { enqRef =>
+                    val modify =
+                      F.modify(qref) { s =>
+                        if (s.enq.nonEmpty) F.pure(registerEnqueue(s, enqRef, xa))
+                        else {
+                          F.bind(fillDequeuer(xa,s.deq)) {
+                            case Right(deq) => F.bind(F.setPure(enqRef)(0))(_ => F.pure(s.copy(queue =Vector.empty, deq = deq)))
+                            case Left(remains) =>
+                              lazy val (queue,rem) = remains.splitAt(bound)
+                              if (bound <= 0 || rem.isEmpty) F.bind(F.setPure(enqRef)(remains.size))(_ => F.pure(s.copy( queue = remains.toVector, deq = Vector.empty)))
+                              else  F.pure(registerEnqueue(s.copy(queue = queue.toVector, deq = Vector.empty),enqRef,rem))
+                          }
+                        }
+                      }
+
+                    F.bind(modify) { case Change(s,ns) =>
+                      F.bind(signalSize(s,ns)) { _ =>
+                        F.get(enqRef)
+                      }
+                    }
+                  }
+                }
+              }
+
+              // implementation of dequeueBatch
+              // repeatedly dequeue from the queue, while completing any waiting enqueue on bounded queue
+              // and taking the chunks up to `limit` that are available in queue
+              def _dequeueBatch(limit: Int): Stream[F, A] = Stream.eval {
```

Comment:
I don't think this function is needed as we discussed before. The consumer of the queue can just use awaitN, awaitLimit, etc to achieve the same thing. No need to make the implementation more complicated than it needs to be.